packages/amplify-provider-awscloudformation/src/utils/consolidate-apigw-policies.ts

```diff
@@ -0,0 +1,223 @@
+    policyDocument: {
+      Version: '2012-10-17',
+      Statement: [{ Effect: 'Allow', Action: ['execute-api:Invoke'], Resource: [] }],
+    },
+  });
+}
+
+function createApiResource(region, account, api, env, method, resourceName) {
+  return cdk.Fn.join('', [
+    'arn:aws:execute-api:',
+    region,
+    ':',
+    account,
+    ':',
+    api,
+    '/',
+    (cdk.Fn.conditionIf('ShouldNotCreateEnvResources', 'Prod', env) as unknown) as string,
+    method,
+    resourceName,
+  ]);
+}
+
+function computePolicySizeIncrease(methodLength: number, pathLength: number, nameLength: number): number {
+  // Each path + HTTP method increases the policy size by roughly:
+  // - 2 * ~190 for the resource boilerplate, etc. (Whitespace is not counted)
+  // - 2 * the length of the HTTP method (ie /POST)
+  // - 2 * the length of the policy resource name (ie /items)
+  // - 2 * the length of the API resourceName
+  return 380 + 2 * (methodLength + pathLength + nameLength);
+}
+
+export async function consolidateApiGatewayPolicies(context: any, stackName: string): Promise<object> {
+  if (!FeatureFlags.getBoolean('restAPI.generateConsolidatedManagedPolicies')) {
+    return {};
+  }
+
+  const apiGateways = [];
+  const { amplifyMeta } = context.amplify.getProjectDetails();
+  const apis = amplifyMeta?.api ?? {};
+
+  Object.keys(apis).forEach(resourceName => {
+    const resource = apis[resourceName];
+    const apiParams = loadApiWithPrivacyParams(context, resourceName, resource);
+
+    if (!apiParams) {
+      return;
+    }
+
+    apiGateways.push({ ...resource, resourceName, params: apiParams });
+  });
+
+  if (apiGateways.length === 0) {
+    return {};
+  }
+
+  return createApiGatewayAuthResources(context, stackName, apiGateways);
+}
+
+async function createApiGatewayAuthResources(context: any, stackName: string, apiGateways: any): Promise<object> {
+  const stack = new ApiGatewayAuthStack(undefined, 'Amplify', {
+    description: 'API Gateway policy stack created using Amplify CLI',
+    stackName,
+    apiGateways,
+  });
+  const cfn = stack.toCloudFormation();
+
+  return {
+    APIGatewayAuthURL: await uploadCfnToS3(context, S3_UPLOAD_PATH, cfn),
+  };
+}
+
+async function uploadCfnToS3(context: any, cfnFile: string, cfnData: object): Promise<string> {
+  const s3 = await S3.getInstance(context);
+  const s3Params = {
+    Body: JSON.stringify(cfnData, null, 2),
+    Key: `amplify-cfn-templates/${cfnFile}`,
+  };
+  const projectBucket = await s3.uploadFile(s3Params);
+
+  return `https://s3.amazonaws.com/${projectBucket}/amplify-cfn-templates/${cfnFile}`;
+}
+
+export function loadApiWithPrivacyParams(context: object, name: string, resource: any): object | undefined {
+  if (resource.providerPlugin !== ProviderName || resource.service !== 'API Gateway') {
+    return;
+  }
+
+  const apiParamsPath = path.join(getResourceDirPath(context, 'api', name), API_PARAMS_FILE);
+  const apiParams = JSON.parse(fs.readFileSync(apiParamsPath, 'utf8'));
```

Comment:
nit: you can use JSONUtilities from amplify-cli-core
```suggestion
  const apiParams = JJSONUtilities.readJson(apiParamsPath);
```